---
title: "Using LimoRhyde for Analyzing Differential Rhythmicity/Expression"
author: "Jordan Singer"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analyzing Circadian Transcriptome Data with LimoRhyde}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = '#>')
```

LimoRhyde establishes a framework for the analysis of how biological rhythms are affected by perturbation.  This is an example use case of LimoRhyde, using a dataset of mouse liver gene expression in WT and a single KO condition, extracting only genes which show evidence of rhythmicity using RAIN, and preparing time-series data for analysis with limma.

## Gathering Experimental Data

In this first stage, we will be gathering experimental data from the [GSE34018](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34018) dataset.

### Load Necessary Packages

```{r, message=FALSE}
library(limorhyde)
library(GEOquery)
library(tidyverse)
library(knitr)
library(limma)
library(rain)
library(doParallel)
library(ggplot2)
```

### Gather dataset from GEOquery

Using getGEO from GEOquery, we download the GSE34018 dataset.
```{r, message=FALSE}
gse = getGEO("GSE34018", GSEMatrix = TRUE)
```

We also gather metadata for the experiment.
```{r, message=FALSE}
columns = c(
    'title',
    'geo_accession',
    'background strain:ch1',
    'genotype/variation:ch1',
    'tissue:ch1')
sampleMetadata = pData(phenoData(gse[[1]]))[,columns]
ZT = as.numeric(sub("_.*", "", sub(".*_ZT", "", sampleMetadata[,'title'])))
sampleMetadata["time"] = ZT
colnames(sampleMetadata) = c('title', 'sample', 'strain', 'genotype', 'tissue', 'time')

kable(sampleMetadata[1:5,])
```

```{r}
sm = sampleMetadata %>%
  mutate(cond = factor(genotype, levels=c('wild-type', 'Reverb alpha/beta DKO')))
```

We can now apply LimoRhyde to our dataset by appending two new columns to the metadata; `time_cos` and `time_sin`, which are generated by computing the first harmonic of a Fourier decomposition on the `time` column.
```{r}
sm = bind_cols(sm, limorhyde(sm, 'time'))
```

Now, we populate the expression set and expression matrix for our dataset.
```{r}
eset = gse$GSE34018_series_matrix.txt.gz
emat = log2(Biobase::exprs(eset[,sm$sample]))
```

## Identifying Rhythmic Genes

We'll be using [RAIN](https://bioconductor.org/packages/release/bioc/html/rain.html) to determine rhythmicity.
```{r, message=FALSE}
sm = arrange(sm, cond, time)  # We sort the metadata, because RAIN expects time points to be sorted.
smGroups = distinct(sm, cond)
rhyRain = foreach(ii=1:nrow(smGroups), .combine=rbind) %do% {
  smNow = sm[sm$cond == smGroups[ii,],]
  tt = smNow$time
  ematNow = emat[,smNow$sample]
  
  r = rle(sort(tt))
  deltat = min(diff(unique(tt)))
  
  msrSeq = tibble(tt = seq(min(tt), max(tt), deltat)) %>%
    full_join(tibble(tt = r$values, n = r$lengths), by='tt') %>%
    mutate(n = ifelse(is.na(n), 0, n))
  
  result = rain(t(ematNow), deltat=deltat, period=24, measure.sequence=msrSeq$n)
  result$geneId = rownames(ematNow)
  result = cbind(result, smGroups[ii,])
  return(result)
}
rownames(rhyRain) = NULL
```

```{r, message=FALSE}
kable(rhyRain[1:5,])
```

## Computing Differential Rhythmicity and Differential Expression

We can now use [limma](https://academic.oup.com/nar/article/43/7/e47/2414268) to analyze the genes which we determined in the previous step to be rhythmic.

### Determine Rhythmic Genes with Cutoff
We'll define some constant cutoff values for filtering out statistically irrelevant data points.
```{r}
qvalRhyCutoff = 0.15
qvalDrCutoff = 0.1
```

```{r}
rhyGenes = rhyRain %>%
		group_by(geneId) %>%
		summarize(pVal = min(pVal)) %>%
		group_by() %>%
		mutate(qval = p.adjust(pVal, method='BH')) %>%
		filter(qval <= qvalRhyCutoff)
```

Calculate differential rhythmicity
```{r, message=FALSE}
design = model.matrix(~ cond * (time_cos + time_sin), data=sm)

fit = lmFit(emat, design)
fit = eBayes(fit, trend=TRUE)
drLimma = topTable(fit, coef=5:6, number=Inf)

drLimma$geneId = rownames(drLimma)
rownames(drLimma) = NULL
drLimma = semi_join(drLimma, rhyGenes, by='geneId')
drLimma$adj.P.Val = p.adjust(drLimma$P.Value, method='BH')

kable(drLimma[1:5,])
```

Calculate differential expression
```{r, message=FALSE}
design = model.matrix(~ cond + time_cos + time_sin, data=sm)

fit = lmFit(emat, design)
fit = eBayes(fit, trend=TRUE)
deLimma = topTable(fit, coef=2, number=Inf)

deLimma$geneId = rownames(deLimma)
rownames(deLimma) = NULL
deLimma = anti_join(deLimma, filter(drLimma, adj.P.Val <= qvalDrCutoff), by='geneId')
deLimma$adj.P.Val = p.adjust(deLimma$P.Value, method='BH')

kable(deLimma[1:5,])
```

## Analyzing limma's Output

Plots of genes shown to be differentially rhythmic or expressive.
```{r, message=FALSE}
ggplot(data = deLimma) +
  geom_point(aes(x = logFC, y = -log10(adj.P.Val)), size = 0.2) + 
  labs(title ="Differential Expression", x = expression(log[2]*'(fold-change)'), y = expression(-log[10]*'('*q[DE]*')'))
```

```{r message=FALSE}
geneIdsNow = deLimma$geneId[1:3]

df = as_tibble(t(emat[geneIdsNow,]))
df$sample = colnames(emat[geneIdsNow,])

df = inner_join(sm, df, by='sample') %>%
	select(sample, cond, time, !!geneIdsNow) %>%
	gather(key='geneSymbol', value='expr', !!geneIdsNow)
df$cond = fct_recode(df$cond, 'wild-type'='wild-type', 'knock-out'='Reverb alpha/beta DKO')

ggplot(df) +
	facet_grid(geneSymbol ~ cond, scales='free_y') +
	geom_point(aes(x=time, y=expr, shape=cond, color=geneSymbol), size=1.5) +
	labs(x='Zeitgeber time (h)', y='Expression (norm.)') +
	scale_shape(solid=FALSE, guide=FALSE) +
	scale_color_brewer(type='qual', palette='Dark2', guide=FALSE) +
	scale_x_continuous(limits=c(0, 24), breaks=seq(0, 24, 6))
```
